File: apps/frontend/lib/api/discovery.ts
/**
 * Job Discovery API (§10.5, design §9).
 *
 * One function per backend endpoint under `/api/v1/discovery`.
 * Follows the same raw-Response pattern as other `lib/api/` modules.
 */

import { apiFetch } from './client';

// -------------------------------------------------------------------------- //
// Types (mirroring apps/backend/app/schemas/discovery.py)
// -------------------------------------------------------------------------- //

export type FetchMode = 'http' | 'stealth';

export interface SearchFilters {
  location?: string | null;
  is_remote?: boolean | null;
  hours_old?: number | null;
  results_wanted?: number | null;
  country_indeed?: string | null;
}

export interface JobListing {
  source: string;
  title: string;
  company: string;
  location: string;
  url: string;
  is_remote?: boolean | null;
  description?: string | null;
  posted_at?: string | null;
  salary?: string | null;
  fingerprint: string;
}

export interface Recommendation {
  listing: JobListing;
  match_score: number;
  partial: boolean;
  matched: string[];
  missing: string[];
}

export interface SourceFailure {
  source: string;
  reason: string;
  kind?: string | null;
}

export interface SearchQuery {
  titles: string[];
  search_string: string;
  seniority?: string | null;
  location?: string | null;
  country_indeed?: string | null;
  degraded: boolean;
}

export interface RecommendRequest {
  resume_id: string;
  filters?: SearchFilters | null;
  force_refresh?: boolean;
}

export interface RecommendResponse {
  recommendations: Recommendation[];
  query: SearchQuery | null;
  degraded: boolean;
  cached: boolean;
  failures: SourceFailure[];
}

export interface TailorRequest {
  resume_id: string;
  listing: JobListing;
}

export interface TailorResponse {
  job_id: string;
  resume_id: string;
}

export interface SiteRecipe {
  id?: number | null;
  user_id: string;
  name: string;
  slug: string;
  base_url: string;
  search_url_template: string;
  schema: Record<string, unknown>;
  fetch_mode: FetchMode;
  enabled: boolean;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface SiteRecipeCreate {
  name: string;
  slug: string;
  base_url: string;
  search_url_template: string;
  schema?: Record<string, unknown>;
  fetch_mode?: FetchMode;
  enabled?: boolean;
}

export interface SiteRecipeUpdate {
  name?: string;
  base_url?: string;
  search_url_template?: string;
  schema?: Record<string, unknown>;
  fetch_mode?: FetchMode;
  enabled?: boolean;
}

// -------------------------------------------------------------------------- //
// API functions
// -------------------------------------------------------------------------- //

const PREFIX = '/api/v1/discovery';

/** POST /discovery/recommend — run discovery and return ranked recommendations. */
export async function postRecommend(
  body: RecommendRequest,
  signal?: AbortSignal
): Promise<RecommendResponse> {
  const res = await apiFetch(`${PREFIX}/recommend`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
    signal,
  });
  if (!res.ok) throw new Error(`Discovery recommend failed: ${res.status}`);
  return res.json();
}

/** GET /discovery/recommend/{resume_id} — last cached recommendations. */
export async function getCachedRecommendations(
  resumeId: string,
  filters?: SearchFilters | null,
  signal?: AbortSignal
): Promise<RecommendResponse> {
  const params = new URLSearchParams();
  if (filters?.location) params.set('location', filters.location);
  if (filters?.is_remote != null) params.set('is_remote', String(filters.is_remote));
  if (filters?.hours_old != null) params.set('hours_old', String(filters.hours_old));
  if (filters?.results_wanted != null) params.set('results_wanted', String(filters.results_wanted));
  if (filters?.country_indeed) params.set('country_indeed', filters.country_indeed);
  const qs = params.toString();
  const url = `${PREFIX}/recommend/${encodeURIComponent(resumeId)}${qs ? `?${qs}` : ''}`;
  const res = await apiFetch(url, { method: 'GET', signal });
  if (!res.ok) throw new Error(`Cached recommendations failed: ${res.status}`);
  return res.json();
}

/** POST /discovery/tailor — hand a listing to the tailor flow. */
export async function postTailor(body: TailorRequest): Promise<TailorResponse> {
  const res = await apiFetch(`${PREFIX}/tailor`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`Tailor handoff failed: ${res.status}`);
  return res.json();
}

/** GET /discovery/recipes — list the user's site recipes. */
export async function listRecipes(signal?: AbortSignal): Promise<SiteRecipe[]> {
  const res = await apiFetch(`${PREFIX}/recipes`, { method: 'GET', signal });
  if (!res.ok) throw new Error(`List recipes failed: ${res.status}`);
  return res.json();
}

/** POST /discovery/recipes — create a new recipe. */
export async function createRecipe(body: SiteRecipeCreate): Promise<SiteRecipe> {
  const res = await apiFetch(`${PREFIX}/recipes`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`Create recipe failed: ${res.status}`);
  return res.json();
}

/** PUT /discovery/recipes/{slug} — update a recipe. */
export async function updateRecipe(slug: string, body: SiteRecipeUpdate): Promise<SiteRecipe> {
  const res = await apiFetch(`${PREFIX}/recipes/${encodeURIComponent(slug)}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`Update recipe failed: ${res.status}`);
  return res.json();
}

/** DELETE /discovery/recipes/{slug} — delete a recipe. */
export async function deleteRecipe(slug: string): Promise<void> {
  const res = await apiFetch(`${PREFIX}/recipes/${encodeURIComponent(slug)}`, {
    method: 'DELETE',
  });
  if (!res.ok) throw new Error(`Delete recipe failed: ${res.status}`);
}


// -------------------------------------------------------------------------- //
// Feed endpoints (Phase 1 — background discovery)
// -------------------------------------------------------------------------- //

export interface FeedResult {
  id: string;
  fingerprint: string;
  source: string;
  title: string;
  company: string;
  location: string;
  url: string;
  is_remote?: boolean | null;
  description?: string | null;
  salary?: string | null;
  posted_at?: string | null;
  match_score: number;
  matched_keywords: string[];
  missing_keywords: string[];
  partial: boolean;
  status: string;
  seen: boolean;
  created_at: string;
  /** The job-description row created when this was saved, if it has been. */
  job_id?: string | null;
  /**
   * Other boards carrying this same job. Present only when duplicates were
   * collapsed, so its absence means "seen once", not "unknown".
   */
  also_on?: string[];
  /** How many board listings collapsed into this row, including this one. */
  duplicate_count?: number;
}

export interface FeedResponse {
  results: FeedResult[];
  total: number;
  /**
   * Rows on this page after same-job duplicates collapsed. Lower than the page
   * size when boards overlapped - reported separately so the page can explain
   * the difference instead of looking like it lost jobs.
   */
  shown?: number;
  unseen: number;
  /**
   * How many jobs in the whole feed carry a real match score. Zero means nothing
   * has been matched against a resume yet, so a score filter could only ever
   * return nothing - the UI hides the control instead of offering a dead end.
   */
  scored: number;
  limit: number;
  offset: number;
}

export interface ScheduleResponse {
  schedule: Record<string, unknown>;
  message: string;
}

/** GET /discovery/feed — paginated job feed. */
/** Filters accepted by the feed. All optional, combined with AND server-side. */
export interface FeedParams {
  status?: string;
  /** Board ids to include. Omit or leave empty for every board. */
  sources?: string[];
  /** Every token must appear in the title or company. */
  q?: string;
  location?: string;
  isRemote?: boolean;
  /** Match percentage floor, 0-100 as the UI shows it. */
  minScore?: number;
  /** Recency window in hours. Jobs with no published date use when we found them. */
  postedWithinHours?: number;
  limit?: number;
  offset?: number;
}

export async function getFeed(
  params?: FeedParams,
  signal?: AbortSignal,
): Promise<FeedResponse> {
  const qs = new URLSearchParams();
  if (params?.status) qs.set('status', params.status);
  if (params?.sources?.length) qs.set('sources', params.sources.join(','));
  if (params?.q?.trim()) qs.set('q', params.q.trim());
  if (params?.location?.trim()) qs.set('location', params.location.trim());
  if (params?.isRemote) qs.set('is_remote', 'true');
  if (params?.minScore) qs.set('min_score', String(params.minScore));
  if (params?.postedWithinHours) qs.set('posted_within_hours', String(params.postedWithinHours));
  if (params?.limit) qs.set('limit', String(params.limit));
  if (params?.offset) qs.set('offset', String(params.offset));
  const query = qs.toString();
  const res = await apiFetch(`${PREFIX}/feed${query ? `?${query}` : ''}`, { method: 'GET', signal });
  if (!res.ok) throw new Error(`Feed failed: ${res.status}`);
  return res.json();
}

/** GET /discovery/feed/unseen — badge count. */
export async function getUnseenCount(signal?: AbortSignal): Promise<{ unseen: number }> {
  const res = await apiFetch(`${PREFIX}/feed/unseen`, { method: 'GET', signal });
  if (!res.ok) throw new Error(`Unseen count failed: ${res.status}`);
  return res.json();
}

/** POST /discovery/feed/schedule — enable background discovery. */
export async function enableSchedule(
  resumeId: string,
  intervalHours: number = 24,
): Promise<ScheduleResponse> {
  const res = await apiFetch(`${PREFIX}/feed/schedule?resume_id=${encodeURIComponent(resumeId)}&interval_hours=${intervalHours}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
  });
  if (!res.ok) throw new Error(`Enable schedule failed: ${res.status}`);
  return res.json();
}

/** POST /discovery/feed/schedule/toggle — pause/resume background discovery. */
export async function toggleSchedule(
  resumeId: string,
  enabled: boolean,
): Promise<{ enabled: boolean; message: string }> {
  const res = await apiFetch(
    `${PREFIX}/feed/schedule/toggle?resume_id=${encodeURIComponent(resumeId)}&enabled=${enabled}`,
    { method: 'POST', headers: { 'Content-Type': 'application/json' } },
  );
  if (!res.ok) throw new Error(`Toggle schedule failed: ${res.status}`);
  return res.json();
}


// -------------------------------------------------------------------------- //
// Manual search (no resume required)
// -------------------------------------------------------------------------- //

export interface ManualSearchRequest {
  query: string;
  location?: string | null;
  is_remote?: boolean | null;
  hours_old?: number | null;
  results_wanted?: number | null;
  country_indeed?: string | null;
  sites?: string[] | null;
  job_type?: string | null; // fulltime, parttime, internship, contract
  distance?: number | null; // miles from location
  resume_id?: string | null;
}

export interface ManualSearchResponse {
  results: Array<{
    listing: JobListing;
    match_score: number;
    partial: boolean;
    matched: string[];
    missing: string[];
  }>;
  total: number;
  query: string;
  sites: string[];
  degraded: boolean;
  failures: Array<{ source: string; reason: string }>;
}

/** POST /discovery/search — manual job search (no resume needed). */
export async function manualSearch(
  body: ManualSearchRequest,
  signal?: AbortSignal,
): Promise<ManualSearchResponse> {
  const res = await apiFetch(`${PREFIX}/search`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
    signal,
  });
  if (!res.ok) throw new Error(`Manual search failed: ${res.status}`);
  return res.json();
}


// -------------------------------------------------------------------------- //
// Status + Cleanup + Schedule editing
// -------------------------------------------------------------------------- //

/** PATCH /discovery/feed/{id}/status — update a job's status. */
export async function updateResultStatus(
  resultId: string,
  status: string,
): Promise<{ id: string; status: string; queued?: boolean }> {
  const res = await apiFetch(`${PREFIX}/feed/${encodeURIComponent(resultId)}/status`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ status }),
  });
  if (!res.ok) throw new Error(`Update status failed: ${res.status}`);
  return res.json();
}

/** POST /discovery/feed/cleanup — archive old results. */
export async function cleanupFeed(days: number = 30): Promise<{ deleted: number }> {
  const res = await apiFetch(`${PREFIX}/feed/cleanup?days=${days}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
  });
  if (!res.ok) throw new Error(`Cleanup failed: ${res.status}`);
  return res.json();
}

/** PATCH /discovery/feed/schedule — edit schedule. */
export async function editSchedule(
  resumeId: string,
  updates: { interval_hours?: number; resume_id?: string },
): Promise<{ message: string }> {
  const res = await apiFetch(`${PREFIX}/feed/schedule?resume_id=${encodeURIComponent(resumeId)}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(updates),
  });
  if (!res.ok) throw new Error(`Edit schedule failed: ${res.status}`);
  return res.json();
}

File: apps/backend/app/schemas/discovery.py
"""HTTP request/response contracts for Job Discovery & Recommendations.

These pydantic models are the wire-level shapes the discovery router
(``routers/discovery.py``) validates and serializes. They are deliberately
separate from the transport-agnostic in-process dataclasses in
:mod:`app.job_discovery.models`; a thin mapping layer in the service converts
between the two.

Design reference: ``.kiro/specs/job-discovery/design.md`` §9 (API surface).
Requirements: 1.4, 4.3, 6.1, 7.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.job_discovery.models import FetchMode

# --------------------------------------------------------------------------- #
# Shared / nested
# --------------------------------------------------------------------------- #


class SearchFiltersIn(BaseModel):
    """User-supplied search constraints (Req 3).

    Every field is optional; the pipeline applies what it understands and
    ignores the rest so a single unknown filter never fails a request.
    """

    model_config = ConfigDict(extra="ignore")

    location: str | None = None
    is_remote: bool | None = None
    hours_old: int | None = Field(default=None, ge=0)
    results_wanted: int | None = Field(default=None, ge=1)
    country_indeed: str | None = None


class JobListingOut(BaseModel):
    """A normalized job posting as returned to the client (design §3.2)."""

    source: str
    title: str
    company: str
    location: str
    url: str
    is_remote: bool | None = None
    description: str | None = None
    posted_at: datetime | None = None
    salary: str | None = None
    fingerprint: str = ""


class RecommendationOut(BaseModel):
    """A ranked listing with match metadata (design §3.2, Req 7)."""

    listing: JobListingOut
    match_score: float = Field(ge=0, le=100)
    partial: bool = False
    matched: list[str] = Field(default_factory=list)
    missing: list[str] = Field(default_factory=list)


class SourceFailureOut(BaseModel):
    """A single non-fatal source failure surfaced to the caller (Req 3.2, 4)."""

    source: str
    reason: str
    kind: str | None = None


class SearchQueryOut(BaseModel):
    """The search intent the pipeline actually used (design §5, Req 2)."""

    titles: list[str] = Field(default_factory=list)
    search_string: str = ""
    seniority: str | None = None
    location: str | None = None
    country_indeed: str | None = None
    degraded: bool = False


# --------------------------------------------------------------------------- #
# Recommend
# --------------------------------------------------------------------------- #


class RecommendRequest(BaseModel):
    """Body for ``POST /discovery/recommend`` (Req 1)."""

    model_config = ConfigDict(extra="ignore")

    resume_id: str
    filters: SearchFiltersIn | None = None
    # Bypass the content-addressed cache and re-run the fan-out.
    force_refresh: bool = False


class RecommendResponse(BaseModel):
    """Response for a recommend / cached-get call (Req 1, 3.2, 7).

    ``degraded`` is True when any source failed or the query was synthesized
    without the LLM, so the client can show a partial-results banner.
    """

    recommendations: list[RecommendationOut] = Field(default_factory=list)
    query: SearchQueryOut | None = None
    degraded: bool = False
    cached: bool = False
    failures: list[SourceFailureOut] = Field(default_factory=list)


# --------------------------------------------------------------------------- #
# Tailor handoff
# --------------------------------------------------------------------------- #


class TailorRequest(BaseModel):
    """Body for the tailor handoff (Req 8): create a job from a listing and
    hand back the ids the tailor flow needs."""

    model_config = ConfigDict(extra="ignore")

    resume_id: str
    listing: JobListingOut


class TailorResponse(BaseModel):
    """Result of the tailor handoff: the created job + the resume to tailor."""

    job_id: str
    resume_id: str


# --------------------------------------------------------------------------- #
# Site recipe CRUD
# --------------------------------------------------------------------------- #


class SiteRecipeCreate(BaseModel):
    """Body for creating a custom-site recipe (Req 4)."""

    model_config = ConfigDict(extra="ignore")

    name: str
    slug: str
    base_url: str
    search_url_template: str
    schema_: dict = Field(default_factory=dict, alias="schema")
    fetch_mode: FetchMode = "http"
    enabled: bool = True


class SiteRecipeUpdate(BaseModel):
    """Body for updating a recipe; all fields optional (partial update)."""

    model_config = ConfigDict(extra="ignore")

    name: str | None = None
    base_url: str | None = None
    search_url_template: str | None = None
    schema_: dict | None = Field(default=None, alias="schema")
    fetch_mode: FetchMode | None = None
    enabled: bool | None = None


class SiteRecipeOut(BaseModel):
    """A persisted recipe as returned to the client (design §3.2, Req 4)."""

    model_config = ConfigDict(populate_by_name=True)

    id: int | None = None
    user_id: str
    name: str
    slug: str
    base_url: str
    search_url_template: str
    schema_: dict = Field(default_factory=dict, serialization_alias="schema")
    fetch_mode: FetchMode = "http"
    enabled: bool = True
    created_at: datetime | None = None
    updated_at: datetime | None = None


__all__ = [
    "SearchFiltersIn",
    "JobListingOut",
    "RecommendationOut",
    "SourceFailureOut",
    "SearchQueryOut",
    "RecommendRequest",
    "RecommendResponse",
    "TailorRequest",
    "TailorResponse",
    "SiteRecipeCreate",
    "SiteRecipeUpdate",
    "SiteRecipeOut",
]
